File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        def BFS(root):
            q=deque([])
            if not root:
                return []
            result=[]
            q.append(root)
            level=0
            while q:
                l=len(q)
                ans=[]
                for i in range(l):
                    r=q.popleft()
                    if r.left:
                        q.append(r.left)
                    if r.right:
                        q.append(r.right)
                    if level%2==0:
                        ans.append(r.val)
                    else:
                        ans.insert(0,r.val)
                result.append(ans) 
                level+=1
            return result
        return BFS(root)

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<List<Integer>> BFS(TreeNode root){
        List<List<Integer>> result=new ArrayList<>();
        if(root==null){
            return result;
        }
        Queue<TreeNode> q=new LinkedList<>();
        q.add(root);
        TreeNode r;
        int l,level=0,i;
        while(!q.isEmpty()){
            l=q.size();
            List<Integer> ans=new ArrayList<>();
            for(i=0;i<l;i++){
               r=q.poll();
               if(r.left!=null){
                    q.add(r.left);
               }
               if(r.right!=null){
                    q.add(r.right);
               }
                if(level%2==0){
                    ans.add(r.val);
                }
                else{
                    ans.add(0,r.val);
                }
            }
            result.add(ans);
            level++;
        }
        return result;
    }
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        return BFS(root);
    }
}